keps/sig-storage/20190530-pv-health-monitor.md

```diff
@@ -0,0 +1,576 @@
+    // GET_VOLUME_HEALTH controller capability is supported.
+    VolumeHealth health = 1;
+  }
+
+  // This field is REQUIRED
+  Volume volume = 1;
+
+  // This field is OPTIONAL. This field MUST be specified if the
+  // GET_VOLUME_HEALTH controller capability is supported.
+  VolumeStatus status = 2;
+}
+```
+
+```
+message VolumeHealth {
+  // The error code describing the health condition of the volume.
+  // This field is REQUIRED.
+  string error_code = 1;
+
+  // The error message associated with the above error_code. This field is OPTIONAL.
+  string message = 2;
+}
+```
+
+The following common error codes are proposed for volume health:
+* VolumeNotFound
+* VolumeUnmounted
+* OutOfCapacity
+* RWIOError
+* DiskFailure
+* DiskRemoved
+* ConfigError
+* NetworkError
+* DiskDegrading
+* FilesystemCorruption
+* NodeFailure
+
+Add `GET_VOLUME` and `GET_VOLUME_HEALTH` controller capabilities.  If a driver supports `GET_VOLUME` capability, it MUST implement the `GetVolume` RPC.  If a driver supports `GET_VOLUME_HEALTH`, it MUST supports the `health` field in the `status` field in `GetVolumeResponse`.
+
+```
+// Specifies a capability of the controller service.
+message ControllerServiceCapability {
+  message RPC {
+    enum Type {
+      UNKNOWN = 0;
+      CREATE_DELETE_VOLUME = 1;
+      PUBLISH_UNPUBLISH_VOLUME = 2;
+      LIST_VOLUMES = 3;
+      GET_CAPACITY = 4;
+      // Currently the only way to consume a snapshot is to create
+      // a volume from it. Therefore plugins supporting
+      // CREATE_DELETE_SNAPSHOT MUST support creating volume from
+      // snapshot.
+      CREATE_DELETE_SNAPSHOT = 5;
+      LIST_SNAPSHOTS = 6;
+
+      // Plugins supporting volume cloning at the storage level MAY
+      // report this capability. The source volume MUST be managed by
+      // the same plugin. Not all volume sources and parameters
+      // combinations MAY work.
+      CLONE_VOLUME = 7;
+
+      // Indicates the SP supports ControllerPublishVolume.readonly
+      // field.
+      PUBLISH_READONLY = 8;
+
+      // See VolumeExpansion for details.
+      EXPAND_VOLUME = 9;
+
+      // Indicates the SP supports the
+      // ListVolumesResponse.entry.published_nodes field
+      LIST_VOLUMES_PUBLISHED_NODES = 10;
+
+      // Indicates the SP supports the GetVolume RPC
+      GET_VOLUME = 11;
+
+      // Indicates the SP supports the
+      // GetVolumeResponse.volume_health field
+      GET_VOLUME_HEALTH = 12;
+    }
+
+    Type type = 1;
+  }
+
+  oneof type {
+    // RPC that the controller supports.
+    RPC rpc = 1;
+  }
+}
+```
+
+#### Add Node Volume Health Function
+
+Node Volume Health checks if a volume is still mounted and usable. To check whether a volume is usable, the CSI driver shall check if filesystem is corrupted, whether there are bad blocks, etc. in this RPC.
+
+Instead of adding a new RPC, we can leverage the existing NodeGetVolumeStats RPC.
+
+```
+rpc NodeGetVolumeStats (NodeGetVolumeStatsRequest)
```

Comment:
Pull based method for fetching utilization makes sense, but for status changes (health) maybe we should consider push?

Reply:
Push based method is brought up in the section "HTTP(RPC) Service".  We didn't adopt that approach as the main proposal because CSI uses pull based method.  I'll add some clarification there.

Reply:
For a push model, can the node agent define some RPC service that CSI drivers can call out to?

@serathius can you point to any existing design that uses a push based model?

Reply:
In the current CSI spec, I don't see a push model.  Therefore the design is based on a pull model.  We can ask about it at the CSI community meeting during the CSI spec review.  If a push model is possible, we can come back to update the KEP later.